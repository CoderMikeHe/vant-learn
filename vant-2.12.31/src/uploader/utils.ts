import { isFunction } from '../utils';

export type ResultType = 'dataUrl' | 'text' | 'file';

export type UploaderMaxSize = number | string | ((file: File) => boolean);

export function toArray<T>(item: T | T[]): T[] {
  if (Array.isArray(item)) {
    return item;
  }

  return [item];
}

export function readFile(file: File, resultType: ResultType) {
  return new Promise<FileReader['result']>((resolve) => {
    if (resultType === 'file') {
      resolve(null);
      return;
    }

    const reader = new FileReader();

    reader.onload = (event) => {
      resolve((event.target as FileReader).result);
    };

    if (resultType === 'dataUrl') {
      reader.readAsDataURL(file);
    } else if (resultType === 'text') {
      reader.readAsText(file);
    }
  });
}

export function isOversize(
  files: File | File[],
  maxSize: UploaderMaxSize
): boolean {
  return toArray(files).some((file) => {
    if (file) {
      if (isFunction(maxSize)) {
        return maxSize(file);
      }
      return file.size > maxSize;
    }
    return false;
  });
}

export type FileListItem = {
  url?: string;
  file?: File;
  content?: string; // dataUrl
  isImage?: boolean;
  status?: '' | 'uploading' | 'done' | 'failed';
  message?: string;
};

const IMAGE_REGEXP = /\.(jpeg|jpg|gif|png|svg|webp|jfif|bmp|dpg)/i;

export function isImageUrl(url: string): boolean {
  return IMAGE_REGEXP.test(url);
}

export function isImageFile(item: FileListItem): boolean {
  // some special urls cannot be recognized
  // user can add `isImage` flag to mark it as an image url
  if (item.isImage) {
    return true;
  }

  if (item.file && item.file.type) {
    return item.file.type.indexOf('image') === 0;
  }

  if (item.url) {
    return isImageUrl(item.url);
  }

  if (item.content) {
    return item.content.indexOf('data:image') === 0;
  }

  return false;
}
