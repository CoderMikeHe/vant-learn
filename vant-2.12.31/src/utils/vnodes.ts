import { VNode } from 'vue';

function flattenVNodes(vnodes: VNode[]) {
  const result: VNode[] = [];

  function traverse(vnodes: VNode[]) {
    vnodes.forEach((vnode) => {
      result.push(vnode);

      // componentInstance: 当前节点对应的组件的实例
      if (vnode.componentInstance) {
        traverse(vnode.componentInstance.$children.map((item) => item.$vnode));
      }

      if (vnode.children) {
        traverse(vnode.children);
      }
    });
  }

  traverse(vnodes);
  return result;
}

// sort children instances by vnodes order
export function sortChildren(children: Vue[], parent: Vue) {

  const { componentOptions } = parent.$vnode;
  // componentOptions: 组件的option选项
  if (!componentOptions || !componentOptions.children) {
    return;
  }
  // 将所有的节点  展开成一维数组
  const vnodes = flattenVNodes(componentOptions.children);
  children.sort((a, b) => vnodes.indexOf(a.$vnode) - vnodes.indexOf(b.$vnode));
}
