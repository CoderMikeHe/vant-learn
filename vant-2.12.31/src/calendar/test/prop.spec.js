import Calendar from '..';
import { mount, later } from '../../../test';
import { minDate, maxDate, formatRange, formatDate } from './utils';

test('max-range prop when type is range and showConfirm is false', async () => {
  const wrapper = mount(Calendar, {
    propsData: {
      type: 'range',
      minDate,
      maxDate,
      maxRange: 3,
      poppable: false,
      showConfirm: false,
    },
  });

  await later();

  const days = wrapper.findAll('.van-calendar__day');
  days.at(12).trigger('click');
  days.at(18).trigger('click');

  expect(formatRange(wrapper.emitted('select')[0][0])).toEqual('2010/1/13-');
  expect(formatRange(wrapper.emitted('select')[1][0])).toEqual(
    '2010/1/13-2010/1/19'
  );
  expect(wrapper.emitted('confirm')).toBeFalsy();
});

test('max-range prop when type is range and showConfirm is true', async () => {
  const wrapper = mount(Calendar, {
    propsData: {
      type: 'range',
      minDate,
      maxDate,
      maxRange: 3,
      poppable: false,
    },
  });

  await later();

  const days = wrapper.findAll('.van-calendar__day');
  days.at(12).trigger('click');
  days.at(18).trigger('click');

  expect(formatRange(wrapper.emitted('select')[0][0])).toEqual('2010/1/13-');
  expect(formatRange(wrapper.emitted('select')[1][0])).toEqual(
    '2010/1/13-2010/1/15'
  );
  expect(wrapper.emitted('confirm')).toBeFalsy();
});

test('max-range prop when type is multiple', async () => {
  const wrapper = mount(Calendar, {
    propsData: {
      type: 'multiple',
      minDate,
      maxDate,
      maxRange: 2,
      poppable: false,
      showConfirm: false,
    },
  });

  await later();

  const days = wrapper.findAll('.van-calendar__day');
  days.at(13).trigger('click');
  days.at(14).trigger('click');

  expect(wrapper.emitted('select').length).toEqual(1);
});

test('show-title prop', () => {
  const wrapper = mount(Calendar, {
    propsData: {
      value: true,
    },
  });

  expect(wrapper.contains('.van-calendar__header-title')).toBeTruthy();
  wrapper.setProps({ showTitle: false });
  expect(wrapper.contains('.van-calendar__header-title')).toBeFalsy();
});

test('show-subtitle prop', () => {
  const wrapper = mount(Calendar, {
    propsData: {
      value: true,
    },
  });

  expect(wrapper.contains('.van-calendar__header-subtitle')).toBeTruthy();
  wrapper.setProps({ showSubtitle: false });
  expect(wrapper.contains('.van-calendar__header-subtitle')).toBeFalsy();
});

test('hide close icon when there is no title', () => {
  const wrapper = mount(Calendar, {
    propsData: {
      value: true,
    },
  });

  expect(wrapper.contains('.van-popup__close-icon')).toBeTruthy();

  wrapper.setProps({
    showTitle: false,
    showSubtitle: false,
  });
  expect(wrapper.contains('.van-popup__close-icon')).toBeFalsy();
});

test('allow-same-day prop', async () => {
  const select = jest.fn();
  const wrapper = mount(Calendar, {
    propsData: {
      type: 'range',
      minDate,
      maxDate,
      poppable: false,
    },
    listeners: {
      select,
    },
  });

  await later();

  const days = wrapper.findAll('.van-calendar__day');
  days.at(9).trigger('click');
  days.at(9).trigger('click');

  expect(select).toHaveBeenLastCalledWith([minDate, null]);

  wrapper.setProps({
    allowSameDay: true,
  });

  days.at(9).trigger('click');
  expect(select).toHaveBeenLastCalledWith([minDate, minDate]);
});

test('min-date after current time', () => {
  const wrapper = mount(Calendar, {
    propsData: {
      poppable: false,
      minDate: new Date(2200, 0, 1),
      maxDate: new Date(2200, 0, 2),
    },
  });

  wrapper.find('.van-calendar__confirm').trigger('click');
  expect(formatDate(wrapper.emitted('confirm')[0][0])).toEqual('2200/1/1');
});

test('min-date before current time', () => {
  const wrapper = mount(Calendar, {
    propsData: {
      poppable: false,
      minDate: new Date(1800, 0, 1),
      maxDate: new Date(1800, 0, 2),
    },
  });

  wrapper.find('.van-calendar__confirm').trigger('click');
  expect(formatDate(wrapper.emitted('confirm')[0][0])).toEqual('1800/1/2');
});

test('lazy-render prop', () => {
  const wrapper = mount(Calendar, {
    propsData: {
      minDate,
      maxDate,
      poppable: false,
      lazyRender: false,
    },
  });

  expect(wrapper).toMatchSnapshot();
});

test('month-show event', async () => {
  const wrapper = mount(Calendar, {
    propsData: {
      value: true,
    },
  });
  await later();

  expect(wrapper.emitted('month-show')).toBeTruthy();
});

test('first day of week', async () => {
  const wrapper = mount(Calendar, {
    propsData: {
      poppable: false,
      defaultDate: new Date(2020, 7, 1),
      minDate: new Date(2020, 7, 1),
      maxDate: new Date(2020, 7, 30),
      firstDayOfWeek: 2,
    },
  });

  await later();

  expect(wrapper.find('.van-calendar__weekdays').text()[0]).toEqual('äºŒ');

  const day = wrapper.find(
    '.van-calendar__month:first-of-type .van-calendar__day'
  );
  expect(day.text()).toEqual('1');
  expect(day.attributes('style')).toContain(`margin-left: ${(100 * 4) / 7}%`);
});

test('readonly prop', async () => {
  const wrapper = mount(Calendar, {
    propsData: {
      type: 'multiple',
      minDate,
      maxDate,
      readonly: true,
      poppable: false,
    },
  });

  await later();

  const days = wrapper.findAll('.van-calendar__day');
  days.at(13).trigger('click');
  expect(wrapper.emitted('select')).toBeFalsy();

  wrapper.setProps({ readonly: false });
  days.at(13).trigger('click');
  expect(wrapper.emitted('select')).toBeTruthy();
});
