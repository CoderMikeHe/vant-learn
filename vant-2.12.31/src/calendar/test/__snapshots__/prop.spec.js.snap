// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lazy-render prop 1`] = `
<div class="van-calendar">
  <div class="van-calendar__header">
    <div class="van-calendar__header-title">日期选择</div>
    <div class="van-calendar__header-subtitle"></div>
    <div class="van-calendar__weekdays"><span class="van-calendar__weekday">日</span><span class="van-calendar__weekday">一</span><span class="van-calendar__weekday">二</span><span class="van-calendar__weekday">三</span><span class="van-calendar__weekday">四</span><span class="van-calendar__weekday">五</span><span class="van-calendar__weekday">六</span></div>
  </div>
  <div class="van-calendar__body">
    <div class="van-calendar__month">
      <div role="grid" class="van-calendar__days">
        <div class="van-calendar__month-mark">1</div>
        <div role="gridcell" class="van-calendar__day van-calendar__day--disabled" style="margin-left: 71.42857142857143%;">1</div>
        <div role="gridcell" class="van-calendar__day van-calendar__day--disabled">2</div>
        <div role="gridcell" class="van-calendar__day van-calendar__day--disabled">3</div>
        <div role="gridcell" class="van-calendar__day van-calendar__day--disabled">4</div>
        <div role="gridcell" class="van-calendar__day van-calendar__day--disabled">5</div>
        <div role="gridcell" class="van-calendar__day van-calendar__day--disabled">6</div>
        <div role="gridcell" class="van-calendar__day van-calendar__day--disabled">7</div>
        <div role="gridcell" class="van-calendar__day van-calendar__day--disabled">8</div>
        <div role="gridcell" class="van-calendar__day van-calendar__day--disabled">9</div>
        <div role="gridcell" tabindex="-1" class="van-calendar__day">10</div>
        <div role="gridcell" tabindex="-1" class="van-calendar__day">11</div>
        <div role="gridcell" tabindex="-1" class="van-calendar__day">12</div>
        <div role="gridcell" tabindex="-1" class="van-calendar__day">13</div>
        <div role="gridcell" tabindex="-1" class="van-calendar__day">14</div>
        <div role="gridcell" tabindex="-1" class="van-calendar__day">15</div>
        <div role="gridcell" tabindex="-1" class="van-calendar__day">16</div>
        <div role="gridcell" tabindex="-1" class="van-calendar__day">17</div>
        <div role="gridcell" tabindex="-1" class="van-calendar__day">18</div>
        <div role="gridcell" tabindex="-1" class="van-calendar__day">19</div>
        <div role="gridcell" tabindex="-1" class="van-calendar__day">
          <div class="van-calendar__selected-day">20</div>
        </div>
        <div role="gridcell" class="van-calendar__day van-calendar__day--disabled">21</div>
        <div role="gridcell" class="van-calendar__day van-calendar__day--disabled">22</div>
        <div role="gridcell" class="van-calendar__day van-calendar__day--disabled">23</div>
        <div role="gridcell" class="van-calendar__day van-calendar__day--disabled">24</div>
        <div role="gridcell" class="van-calendar__day van-calendar__day--disabled">25</div>
        <div role="gridcell" class="van-calendar__day van-calendar__day--disabled">26</div>
        <div role="gridcell" class="van-calendar__day van-calendar__day--disabled">27</div>
        <div role="gridcell" class="van-calendar__day van-calendar__day--disabled">28</div>
        <div role="gridcell" class="van-calendar__day van-calendar__day--disabled">29</div>
        <div role="gridcell" class="van-calendar__day van-calendar__day--disabled">30</div>
        <div role="gridcell" class="van-calendar__day van-calendar__day--disabled">31</div>
      </div>
    </div>
  </div>
  <div class="van-calendar__footer"><button type="button" class="van-button van-button--danger van-button--normal van-button--block van-button--round van-calendar__confirm">
      <div class="van-button__content"><span class="van-button__text">确定</span></div>
    </button></div>
</div>
`;
